server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -238,10 +239,20 @@ class AsyncAction {
                 assert request.concreteNodes() != null;
             }
             this.responses = new AtomicReferenceArray<>(request.concreteNodes().length);
+
+            if (request.retainDiscoveryNodes() == false) {
+                // We transfer the ownership of discovery nodes to route the request to into the AsyncAction class.
+                // This reduces the payload of the request and improves the number of concrete nodes in the memory
+                this.concreteNodes = request.concreteNodes();
+                request.setConcreteNodes(null);
+            } else {
+                // initializing it separately as we keep the `concreteNodes` as final since we want it to be immutable.
+                this.concreteNodes = null;
+            }
```

Comment:
With the current approach, if we decide to retain `DiscoveryNodes`  in the request, we rely on DiscoveryNodes in the request in the AsyncAction.start(). Otherwise if we do not want to retain, we set the concreteNodes on the AsyncAction class and set the request's concreteNodes to `null`. 

This way we reduce the number of DiscoveryNodes in the heap as well as reduce the size of the request being sent.